### PR TITLE
Replace react-waypoint with react-intersection-observer

### PR DESCRIPTION
**Pull request by GeKorm**
**What**:
Replaced react-waypoint with [react-intersection-observer](https://github.com/thebuilder/react-intersection-observer)

**Why**:
The intersection observer might be more suited to the task and is more lightweight, especially when supported natively (currently Edge, Firefox and Chrome).

**How**:
- In the code the only thing that changed is using onEnter/onLeave inside the `handleChange` handler of react-intersection-observer.
- I added a section about polyfilling IntersectionObserver in the README.

I also made a couple "optional" changes to the package json:
- Defined `react-intersection-observer` as a **dependency** instead of a **peerDependency**
- I added [canvas-prebuilt](https://www.npmjs.com/package/canvas-prebuilt) as I couldn't run the tests without it.

**Checklist**:

* [x] Documentation
* [x] Tests
* [x] Ready to be merged 
* [x] Added myself to contributors table

### CODE REVIEW

**Comment by stereobooster** [comment]
I'm not sure about this. `react-waypoint` is reliable and well supported component. I don't want to switch to some other less familiar component (which also requires polyfills). What is the actual win using one over another, do you have any comparison?

**Comment by GeKorm** [comment]
Hi thanks for taking the time. 

Intersection observers handle more cases such as CSS transforms, not just scroll position.  
It should also be generally faster. I can't vouch for the performance of react-interesection-observer's implementation however. We've been using it in production for almost a year but I have no benchmarks vs waypoint. It's a fairly thin wrapper over the native API though.

Thanks again

File: src/components/IdealImage/index.js [hunk 1]
```diff
@@ -309,11 +309,19 @@ export default class IdealImage extends Component {
...
+      this.onLeave()
+    }
+  }
+
   render() {
     const icon = this.props.getIcon(this.state)
     const message = this.props.getMessage(icon, this.state)
     return (
-      <Waypoint onEnter={this.onEnter} onLeave={this.onLeave}>
+      <Observer oChange={this.handleChange}>
```
Comment (nakleiderer) [inline src/components/IdealImage/index.js]: Should `oChange` be `onChange` on this line?

Comment (GeKorm) [reply src/components/IdealImage/index.js]: Woops, pushing a fix thanks

**Comment by disintegrator** [comment]
I'd suggest using IntersectionObserver as is and not pulling in a dependency (or listing a peer dep). As for older browsers or ones with partial support (older versions of Edge), I recommend instructing users to apply the polyfill in the README.

**Comment by stereobooster** [comment]
Based on w3c/IntersectionObserver#258, intersection observer doesn't report in case of programmatic scrolling, but I haven't tested it myself. If this is true it can happen we will need both. Or maybe provide a way to decide by developer which implementation to use. I haven't got a time to investigate it

**Comment by GeKorm** [comment]
I ran the demo on Chrome 67 and Firefox 61 on windows and it's indeed the case. That's super disappointing. Providing a choice might be a good compromise, I'll try to cook something up.

**Comment by disintegrator** [comment]
I understand. I didn't account for programatic scrolling. I found it problematic that react-waypoint adds 10KB of minified code. These days I'm trying to find any way I can keep my main bundle size down and the easiest approach is to rely on native features like IntersectionObserver.

**Comment by hipstersmoothie** [comment]
yeah that extra 10kb is really a bummer @disintegrator @stereobooster

**Comment by stereobooster** [comment]
Solution here is to implement lightweight component with interface similar to waypoint (e.g. onLeave onEnter) and let developer to choose which one to use with Waypoint as default option

**Comment by hipstersmoothie** [comment]
I haven't used this one but it looks tiny and shake-able. https://bundlephobia.com/result?p=react-page-visibility@3.0.0

**Comment by GeKorm** [comment]
@hipstersmoothie That API is for detecting if a browser tab is visible, it's not for individual elements.

**Comment by hipstersmoothie** [comment]
@stereobooster how about we allow the user to pass in a waypoint. And the current `Waypoint` Can be the default.

```js
  <IdealImage
	waypoint={AnyWaypointIWant}
    placeholder={{lqip}}
    srcSet={[{src: image, width: 3500}]}
    alt="doggo"
    width={3500}
    height={2095}
  />
```

**Comment by stereobooster** [comment]
> @stereobooster how about we allow the user to pass in a waypoint. And the current Waypoint Can be the default.

yes that is exactly what I meant, except I wouldn't name property `waypoint`, maybe `visibilityObserver` or something

**Comment by hipstersmoothie** [comment]
To do this I think I need to:

* Make a IdealImageFactory a function the takes a visibilityObserver
* Make the default use Waypoint (I'm not sure how the dependency will work if not installed)

**Comment by stereobooster** [comment]
> Make a IdealImageFactory a function the takes a visibilityObserver

It is a bit simpler

components/IdealImage/index.js
```js
 render() {
    //...
    const { visibilityObserver } = props
    return (
      <visibilityObserver onEnter={this.onEnter} onLeave={this.onLeave}>
```

components/IdealImageWithDefaults/index.js
```js
IdealImageWithDefaults.defaultProps = {
  ...IdealImage.defaultProps,
  visibilityObserver: Waypoint
}
...

**Comment by stereobooster** [comment]
I wonder if this would be enough:

```js
import Observer from 'react-intersection-observer'

const WaypointLike = function({onEnter, onLeave, children}) => (
  <Observer onChange={(inView) => inView ? onEnter() : onLeave() }>{children}</Observer>
)
```

**Comment by disintegrator** [comment]
_This is a rough idea but I think something similar to it might work._

If you use a render prop API you can more intuitively hand over some control to the consumer.

```
<IdealImage

  delayLoad

  placeholder={{lqip}}
  srcSet={[{src: image, width: 3500}]}
  alt="doggo"
  width={3500}
  height={2095}>
  {({ Image, load }) => (
    <Waypoint onEnter={load}>
    	<Image />
    </Waypoint>
  )}
</IdealImage>
```
```
<IdealImage

  delayLoad

  placeholder={{lqip}}
  srcSet={[{src: image, width: 3500}]}
  alt="doggo"
  width={3500}
  height={2095}>
  {({ Image, load }) => (
    <IntersectionObserver onEnter={load}>
    	<Image />
    </IntersectionObserver>
  )}
</IdealImage>
```
When `delayLoad` is set, `load` needs to be called to trigger loading of the actual image.
When `delayLoad` is not set, `load` is a no-op.
`Image` boxes up the current state and renders it.

**Comment by stereobooster** [comment]
Implemented this one https://github.com/stereobooster/react-ideal-image/pull/51 (but haven't tested it). Take a look

**Pull request closed without merging**